pkg/bloomtokenizer/bloomtokenizer.go

```diff
@@ -0,0 +1,98 @@
+package bloomtokenizer
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/loki/pkg/chunkenc"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logql/log"
+	"github.com/grafana/loki/pkg/storage/bloom/v1/filter"
+	"github.com/grafana/loki/pkg/storage/chunk"
+	util_log "github.com/grafana/loki/pkg/util/log"
+	"github.com/grafana/loki/tools/tsdb/helpers"
+)
+
+type metrics struct{}
+
+type BloomTokenizer struct {
+	metrics *metrics
+
+	lineTokenizer    Tokenizer
+	chunkIDTokenizer *WrappedTokenizer
+	cache            map[string]interface{}
+}
+
+const CacheSize = 150000
+
+// New returns a new instance of the Bloom Tokenizer.
+func NewBloomTokenizer(reg prometheus.Registerer) (*BloomTokenizer, error) {
+	t := &BloomTokenizer{
+		metrics: newMetrics(reg),
+	}
+	t.cache = make(map[string]interface{}, CacheSize)
+	// TODO: make these configurable
+	t.lineTokenizer = newNGramTokenizer(4, 5, 0)
+	t.chunkIDTokenizer = ChunkIDTokenizer(t.lineTokenizer)
+
+	level.Info(util_log.Logger).Log("bloom tokenizer created")
+
+	return t, nil
+}
+
+// TODO: Something real here with metrics
+func newMetrics(r prometheus.Registerer) *metrics {
+	return &metrics{}
+}
+
+func clearCache(cache map[string]interface{}) {
```

Comment:
For future improvement: we can version the cache so clearing it doesn't require iterating all the keys. It could be something like `map[key]int` where `int` is the version number and if a key exists but the version isn't the current expected version, treat it as absent

Reply:
I understand where you're going with that, but I believe it will lead to the map growing unbounded.  For an extreme example, if the map was filled to capacity with a set of tokens with version=1, when we bump to version 2, if the next set of tokens were all unique, we'd have capacity*2 number of tokens.
I don't like iterating through the keys, but it saves reallocation of the map and has proven faster in my benchmarks